File: framec_tests/cpp/Event_Handler/Event_Handler.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <unordered_map>
#include <stdexcept>
#include <string>
#include <iostream>
#include <vector>
#include <any>
using namespace std;
#include "../FrameLang/FrameLang.h"

//=============== Compartment ==============//

class EventHandlerCompartment
{
public:
    int state;
    
    EventHandlerCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class EventHandler
{
private:
    EventHandlerCompartment *_compartment_;
    EventHandlerCompartment *_nextCompartment_;
    
    
    
public:
    EventHandler()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(EventHandlerState::S1);
        
        _compartment_ = new EventHandlerCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class EventHandlerState
    {
        S1 = 0,
        S2 = 1
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(EventHandlerState::S1))
        {
            this->_sS1_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(EventHandlerState::S2))
        {
            this->_sS2_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            EventHandlerCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void LogIt(int x) {
            unordered_map<string, any> params;
            params["x"] = x;

            FrameEvent e("LogIt", params);
            this->_mux_(&e);
        }
        
        void LogAdd(int a,int b) {
            unordered_map<string, any> params;
            params["a"] = a;

            // FrameEvent e("LogAdd", params);
            // this->_mux_(&e);
            params["b"] = b;

            FrameEvent e("LogAdd", params);
            this->_mux_(&e);
        }
        
        int LogReturn(int a,int b) {
            unordered_map<string, any> params;
            params["a"] = a;

            // FrameEvent e("LogReturn", params);
            // this->_mux_(&e);
            params["b"] = b;

            FrameEvent e("LogReturn", params);
            this->_mux_(&e);
            return any_cast<int>(e._return);
        }
        
        void PassAdd(int a,int b) {
            unordered_map<string, any> params;
            params["a"] = a;

            // FrameEvent e("PassAdd", params);
            // this->_mux_(&e);
            params["b"] = b;

            FrameEvent e("PassAdd", params);
            this->_mux_(&e);
        }
        
        int PassReturn(int a,int b) {
            unordered_map<string, any> params;
            params["a"] = a;

            // FrameEvent e("PassReturn", params);
            // this->_mux_(&e);
            params["b"] = b;

            FrameEvent e("PassReturn", params);
            this->_mux_(&e);
            return any_cast<int>(e._return);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sS1_(FrameEvent *e)
    {
        if (e->_message == "LogIt") {
            log_do("x",any_cast<int>(e->_parameters["x"]));
            return;
        }
        else if (e->_message == "LogAdd") {
            log_do("a",any_cast<int>(e->_parameters["a"]));
            log_do("b",any_cast<int>(e->_parameters["b"]));
            log_do("a+b",any_cast<int>(e->_parameters["a"]) + any_cast<int>(e->_parameters["b"]));
            return;
        }
        else if (e->_message == "LogReturn") {
            log_do("a",any_cast<int>(e->_parameters["a"]));
            log_do("b",any_cast<int>(e->_parameters["b"]));
            int r  = any_cast<int>(e->_parameters["a"]) + any_cast<int>(e->_parameters["b"]);
            log_do("r",r);
            e->_return = (void*) new int(r);
            
            return;
            
        }
        else if (e->_message == "PassAdd") {
            EventHandlerCompartment *compartment =  new EventHandlerCompartment(static_cast<int>(EventHandlerState::S2));
            compartment->stateArgs["p"] = any_cast<int>(e->_parameters["a"]) + any_cast<int>(e->_parameters["b"]);
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "PassReturn") {
            int r  = any_cast<int>(e->_parameters["a"]) + any_cast<int>(e->_parameters["b"]);
            log_do("r",r);
            EventHandlerCompartment *compartment =  new EventHandlerCompartment(static_cast<int>(EventHandlerState::S2));
            compartment->stateArgs["p"] = r;
            
            this->_transition_(compartment);
            e->_return = (void*) new int(r);
            
            return;
            
        }
    }
    
    void _sS2_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do("p",any_cast<int>(this->_compartment_->stateArgs["p"]));
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    // Unimplemented Actions
    
    // void log_do(std::string msg, int val) {  throw std::logic_error("Not implemented");  }

    public:
    void log_do(std::string msg, int val) {
        std::string value = msg + "=" + to_string(val); 
		tape.push_back(value);
	}
    
    //===================== Domain Block ===================//
    
    
    public:
    std::vector<std::string> tape ;
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(EventHandlerCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(EventHandlerCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class EventHandlerController : public EventHandler
{
public:
	EventHandlerController() : EventHandler() {}
std::string msg, int val) {}
};

********************/

File: framec_tests/javascript/src/event_handler/event_handler.js
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class EventHandler {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sS1_;
        this.#compartment = new EventHandlerCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    LogIt(x) {
        let e = FrameEvent("LogIt",{"x":x});
        this.#mux(e);
    }
    
    LogAdd(a,b) {
        let e = FrameEvent("LogAdd",{"a":a,"b":b});
        this.#mux(e);
    }
    
    LogReturn(a,b) {
        let e = FrameEvent("LogReturn",{"a":a,"b":b});
        this.#mux(e);
        return e._return
    }
    
    PassAdd(a,b) {
        let e = FrameEvent("PassAdd",{"a":a,"b":b});
        this.#mux(e);
    }
    
    PassReturn(a,b) {
        let e = FrameEvent("PassReturn",{"a":a,"b":b});
        this.#mux(e);
        return e._return
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sS1_:
                this.#sS1_(e);
                break;
            case this.#sS2_:
                this.#sS2_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sS1_(e) {
        switch (e._message) {
            case "LogIt":
                {
                this.log_do("x",(e._parameters["x"]));
                
                return;
                }
                
            case "LogAdd":
                {
                this.log_do("a",(e._parameters["a"]));
                this.log_do("b",(e._parameters["b"]));
                this.log_do("a+b",(e._parameters["a"]) + (e._parameters["b"]));
                
                return;
                }
                
            case "LogReturn":
                {
                this.log_do("a",(e._parameters["a"]));
                this.log_do("b",(e._parameters["b"]));
                let r = e._parameters["a"] + e._parameters["b"];
                this.log_do("r",r);
                let compartment =  new EventHandlerCompartment(this.#sS2_);
                
                
                this.#transition(compartment);
                e._return = r
                return;
                }
                
            case "PassAdd":
                {
                let compartment =  new EventHandlerCompartment(this.#sS2_);
                
                compartment.StateArgs["p"] = e._parameters["a"] + e._parameters["b"];
                
                this.#transition(compartment);
                
                return;
                }
                
            case "PassReturn":
                {
                let r = e._parameters["a"] + e._parameters["b"];
                this.log_do("r",r);
                let compartment =  new EventHandlerCompartment(this.#sS2_);
                
                compartment.StateArgs["p"] = r;
                
                this.#transition(compartment);
                e._return = r
                return;
                }
                
        }
    }
    
    #sS2_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("p",(this.#compartment.StateArgs["p"]));
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    log_do(msg,val) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
    
};

//=============== Compartment ==============//

class EventHandlerCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class EventHandlerController extends EventHandler {

	constructor() {
	  super()
	}
	log_do(msg,val) {}
};

********************/

module.exports = EventHandler
